internal/manager/telemetry/telemetry_test.go

```diff
@@ -0,0 +1,446 @@
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1alpha2.TCPRouteSpec{
+						Rules: []gatewayv1alpha2.TCPRouteRule{{}},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+
+			_, err = gcl.GatewayV1alpha2().UDPRoutes(namespace).Create(
+				ctx,
+				&gatewayv1alpha2.UDPRoute{
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1alpha2.UDPRouteSpec{
+						Rules: []gatewayv1alpha2.UDPRouteRule{{}},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+
+			_, err = gcl.GatewayV1alpha2().TLSRoutes(namespace).Create(
+				ctx,
+				&gatewayv1alpha2.TLSRoute{
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1alpha2.TLSRouteSpec{
+						Rules: []gatewayv1alpha2.TLSRouteRule{{}},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+
+			_, err = gcl.GatewayV1beta1().ReferenceGrants(namespace).Create(
+				ctx,
+				&gatewayv1beta1.ReferenceGrant{
+					ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("test-%d", i)},
+					Spec: gatewayv1beta1.ReferenceGrantSpec{
+						From: []gatewayv1beta1.ReferenceGrantFrom{
+							{
+								Kind:      "test",
+								Namespace: metav1.NamespaceDefault,
+							},
+						},
+						To: []gatewayv1beta1.ReferenceGrantTo{
+							{
+								Kind: "test",
+							},
+						},
+					},
+				},
+				metav1.CreateOptions{},
+			)
+			require.NoError(t, err)
+		}
+	}
+}
+
+func handleConnectionToTelemetryServer(ctx context.Context, t *testing.T, listener net.Listener, reportChan chan<- []byte) {
+	// Any function return indicates that either the
+	// report was sent or there was nothing to send.
+	defer close(reportChan)
+	conn, err := listener.Accept()
+	if !assert.NoError(t, err) {
+		return
+	}
+	defer conn.Close()
+	for {
+		report := make([]byte, 2048) // Report is much shorter.
+		n, err := conn.Read(report)
+		if !assert.NoError(t, err) {
+			return
+		}
+		select {
+		case reportChan <- report[:n]:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func verifyTelemetryReport(t *testing.T, k8sVersion *version.Info, report []byte) bool {
+	t.Helper()
+	hostname, err := os.Hostname()
+	if err != nil {
+		t.Logf("failed to get hostname: %s", err)
+		return false
+	}
+	semver, err := utilversion.ParseGeneric(k8sVersion.GitVersion)
+	if err != nil {
+		t.Logf("failed to parse k8s version: %s", err)
+		return false
+	}
+
+	// Report contains stanza like:
+	// id=57a7a76c-25d0-4394-ab9a-954f7190e39a;
+	// uptime=9;
+	// that is not stable across runs, so we need to remove it.
+	actualReport, err := removeStanzaFromReport(string(report), "id")
+	if err != nil {
+		t.Logf("failed to remove stanza id from report: %s", err)
+		return false
+	}
+	actualReport, err = removeStanzaFromReport(actualReport, "uptime")
+	if err != nil {
+		t.Logf("failed to remove stanza uptime from report: %s", err)
+		return false
+	}
```

Comment:
Nit: wouldn't it be better to remove those 2 in a loop, looping over `"id"` and `"uptime"`?